autocomplete/complete.go

```diff
@@ -94,64 +104,188 @@ func getPotentialTarget(prefix string) ([]string, error) {
+	}
+
+	lPath := strings.Split(u.HomeDir, "/")
+	if len(lPath) < 2 {
+		return nil, errors.New("users not found")
 	}
-	currentUser, err := user.Current()
+
+	directoryList, err := ioutil.ReadDir(fmt.Sprintf("/%s", lPath[1]))
+	if err != nil {
+		return nil, err
+	}
+	var res []string
+	for _, s := range directoryList {
+		if !s.IsDir() {
+			continue
+		}
+		u, err := user.Lookup(s.Name())
+		if err != nil {
+			continue
+		}
+		res = append(res, fmt.Sprintf("~%s", u.Username))
+	}
+	return res, nil
+}
+
+func matchAndFilterUser(users []string, prefix string) (match bool, filter []string) {
+	for _, s := range users {
+		if s == prefix {
+			match = true
+			return
+		}
+		if strings.HasPrefix(s, prefix) {
+			filter = append(filter, s)
+		}
+	}
+	return
+}
+
+func resolveUserPath(prefix string) ([]string, error) {
+	users, err := getUsers()
 	if err != nil {
 		return nil, err
 	}
 
-	// TODO expand this logic to support other users
 	if prefix == "~" {
-		prefix = "~/"
+		return users, nil
 	}
 
-	expandedHomeLen := 0
-	if strings.HasPrefix(prefix, "~") {
-		expandedHomeLen = len(currentUser.HomeDir) + 1
-		if len(prefix) > 2 {
-			prefix = currentUser.HomeDir + "/" + prefix[2:]
-		} else {
-			prefix = currentUser.HomeDir + "/"
+	replacer := ""
+	username := ""
+	if strings.HasPrefix(prefix, "~/") {
+		replacer = "~"
+		u, err := user.Current()
+		if err != nil {
+			return nil, err
+		}
+		username = fmt.Sprintf("~%s", u.Username)
+	} else {
+		// match and filter
+		sl := strings.Split(prefix, "/")
+		username = sl[0]
+		replacer = username
+		match, filters := matchAndFilterUser(users, username)
+		// filter users by prefix
+		if !match {
+			return filters, nil
 		}
 	}
 
-	dir, f := path.Split(prefix)
+	u, err := getUser(username)
+	if err != nil {
+		return nil, err
+	}
+
+	if strings.HasPrefix(prefix, "~/") {
+		prefix = strings.Replace(prefix, "~", u.HomeDir, 1)
+	} else {
+		prefix = strings.Replace(prefix, username, u.HomeDir, 1)
+	}
+	var result []string
+	if hasEarthfile(prefix) {
+		x := strings.Replace(prefix, u.HomeDir, replacer, 1)
+		result = append(result, fmt.Sprintf("%s+", x))
+	}
+
+	paths, err := lsPath(prefix)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, p := range paths {
+		result = append(result, strings.Replace(p, u.HomeDir, replacer, 1))
+	}
+	return result, nil
+}
+
+func filterPath(p, filter string) ([]string, error) {
+	fi, err := os.Stat(p)
+	if err != nil {
+		return nil, err
+	}
+	if !fi.IsDir() {
+		return nil, errors.New("not a directory")
+	}
 
-	replaceHomePrefix := func(s string) string {
-		if expandedHomeLen == 0 {
-			return s
+	lDir, err := printDirectory(p)
+	if err != nil {
+		return nil, err
+	}
+	var res []string
+	for _, s := range lDir {
+		if strings.HasPrefix(s, path.Join(p, filter)) {
+			res = append(res, s)
 		}
-		return "~/" + s[expandedHomeLen:]
 	}
+	return res, nil
+}
 
-	files, err := ioutil.ReadDir(dir)
+func printDirectory(prefix string) ([]string, error) {
```

Comment:
nitpick, this function doesn't print anything; perhaps it should be named something like `listDirectory` or `getPotentialDirectoryCompletions` if you're not into the whole brevity thing?

Reply:
yes make sense I changed to listDirectory